custom_components/electrasmart/climate.py

```diff
@@ -214,6 +225,67 @@ def hvac_modes(self):
             HVAC_MODE_HEAT_COOL,
         ]
 
+    @property
+    def preset_modes(self):
+        """PRESET modes."""
+        return [
+            PRESET_NONE,
+            PRESET_SHABAT,
+            PRESET_SLEEP,
+            PRESET_SHABAT_SLEEP,
+            PRESET_IFEEL,
+            PRESET_SHABAT_IFEEL,
+            PRESET_SLEEP_IFEEL,
+            PRESET_SHABAT_SLEEP_IFEEL,
+        ]
+
+    PRESET_MODE_ID = {
+        PRESET_NONE: 0,
+        PRESET_SHABAT: 1,
+        PRESET_SLEEP: 2,
+        PRESET_SHABAT_SLEEP: 3,
+        PRESET_IFEEL: 4,
+        PRESET_SHABAT_IFEEL: 5,
+        PRESET_SLEEP_IFEEL: 6,
+        PRESET_SHABAT_SLEEP_IFEEL: 7,
+    }
+
+    PRESET_MODE_ID_INV = {v: k for k, v in PRESET_MODE_ID.items()}
+
+    PRESET_MODE_MAPPING = {
+        "SHABAT": PRESET_SHABAT,
+        "SLEEP": PRESET_SLEEP,
+        "IFEEL": PRESET_IFEEL,
+    }
+
+    PRESET_MODE_MAPPING_ARGS = {
+        PRESET_SHABAT: "shabat",
+        PRESET_SLEEP: "ac_sleep",
+        PRESET_IFEEL: "ifeel",
+    }
+
+    PRESET_MODE_MAPPING_INV = {v: k for k, v in PRESET_MODE_MAPPING.items()}
+
+    PRESET_MODE_MAPPING_ATTR = {p: p.lower() for p in PRESET_MODE_MAPPING.keys()}
+
+    @property
+    def preset_mode(self):
+        """Returns the current preset mode"""
+        if self.ac.status is None:
+            _LOGGER.debug(f"preset_mode: status is None, returning None")
+            return None
+        if self.ac.status.is_on:
+            preset_mode_bit = 0
+            for p in self.PRESET_MODE_MAPPING.keys():
+                if getattr(self.ac.status, self.PRESET_MODE_MAPPING_ATTR[p]) == "ON":
+                    preset_mode_bit += self.PRESET_MODE_ID[self.PRESET_MODE_MAPPING[p]]
```

Comment:
Best way found to write it at 3am
Implementing a bit field with 3 presets, kind of like RWX